src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,244 @@
+        /// </summary>
+        public bool SkipUnrecognizedProjects { get; set; } = true;
+
+        /// <summary>
+        /// Associates a project file extension with a language name.
+        /// </summary>
+        public void AssociateFileExtensionWithLanguage(string projectFileExtension, string language)
+        {
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            if (projectFileExtension == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileExtension));
+            }
+
+            _projectFileLoaderRegistry.AssociateFileExtensionWithLanguage(projectFileExtension, language);
+        }
+
+        private void SetSolutionProperties(string solutionFilePath)
+        {
+            const string SolutionDirProperty = "SolutionDir";
+
+            // When MSBuild is building an individual project, it doesn't define $(SolutionDir).
+            // However when building an .sln file, or when working inside Visual Studio,
+            // $(SolutionDir) is defined to be the directory where the .sln file is located.
+            // Some projects out there rely on $(SolutionDir) being set (although the best practice is to
+            // use MSBuildProjectDirectory which is always defined).
+            if (!string.IsNullOrEmpty(solutionFilePath))
+            {
+                var solutionDirectory = Path.GetDirectoryName(solutionFilePath);
+                if (!solutionDirectory.EndsWith(@"\", StringComparison.Ordinal))
+                {
+                    solutionDirectory += @"\";
+                }
+
+                if (Directory.Exists(solutionDirectory))
+                {
+                    _properties = _properties.SetItem(SolutionDirProperty, solutionDirectory);
+                }
+            }
+        }
+
+        private DiagnosticReportingMode GetReportingModeForUnrecognizedProjects()
+            => this.SkipUnrecognizedProjects
+                ? DiagnosticReportingMode.Log
+                : DiagnosticReportingMode.Throw;
+
+        /// <summary>
+        /// Loads the <see cref="SolutionInfo"/> for the specified solution file, including all projects referenced by the solution file and 
+        /// all the projects referenced by the project files.
+        /// </summary>
+        public async Task<SolutionInfo> LoadSolutionInfoAsync(
+            string solutionFilePath,
+            IProgress<ProjectLoadProgress> progress = null,
+            CancellationToken cancellationToken = default)
+        {
+            if (solutionFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(solutionFilePath));
+            }
+
+            if (!_pathResolver.TryGetAbsoluteSolutionPath(solutionFilePath, baseDirectory: Directory.GetCurrentDirectory(), DiagnosticReportingMode.Throw, out var absoluteSolutionPath))
```

Comment:
>baseDirectory: Directory.GetCurrentDirectory() [](start = 76, length = 46)

Think Directory.GetCurrentDirectory is fine as a default but feel like this is a value I should be able to specify.

Reply:
This is existing behavior, so I left it. Most of the time, I expect solutionFilePath to be a full path.

Reply:
FWIW, I'm not planning to modify the API to add a `baseDirectory` parameter. In general, the user should pass a full path. If they pass a relative path, I think it's entirely reasonable to expect that the user will have set the CWD. After all, if they want to pass a relative path and a base directory, why wouldn't they just call `Path.Combine(...)` before calling the API.

Reply:
I'll go ahead and add XML doc comments for the parameters to these APIs so that users know what's legal to pass.